### ISSUE TITLE
Animate polyline plugin?

### ISSUE THREAD

**Issue by 1075dn** [fleaflet/flutter_map#897]
Hey everyone,

I'm working on an orienteering app. When the user completes an orienteering course, he'll be shown the route he took. The thing is, I want to be able to animate that route. Anybody interested in collaborating on plugin to animate polylines based on the [drawing_animation](https://pub.dev/packages/drawing_animation) package? (flutter_map uses `Path`s for polylines, which is what the drawing_animation package animates)

Thanks!

**Comment by ibrierley** [comment]
I'd certainly be happy to follow it and give any thoughts. If you want something simple, I assume you just need to interpolate between 0.0 and path.computeMetrics().length, and then draw metric.extractPath(0.0, interpolatedValue) or something around that area.

**Comment by ibrierley** [comment]
In fact, do you need any more than that ? Wrap it in a timer and call setState() on it

**Comment by ibrierley** [comment]
No problem, any issues just shout. I've done a bit of line stuff with SVG, so some of those bits aren't too unfamiliar, but there are often odd differences that can throw a spanner in the works.

**Comment by 1075dn** [comment]
I wrote some working code (based on the existing polyline code, the tappable_polyline plugin, and the drawing_animation package) here: [flutter_map_animated_polyline](https://github.com/1075dn/flutter_map_animated_polyline).
Would you mind checking it out? It's quite messy - if you have any suggestions for how to optimize it or clean it up please let me know!
Thanks again!

**Comment by 1075dn** [comment]
There should be a button in the AppBar (possibly hidden by the debug label) that should animate the polyline.

**Comment by 1075dn** [comment]
Thanks! Which branch were you looking at? I'd advise you to look at [the v2 branch](https://github.com/1075dn/flutter_map_animated_polyline/tree/v2), because it's cleaner and works better (besides for the fact that it doesn't repaint when the interpolation value changes)...

**Comment by 1075dn** [comment]
@ibrierley Did you get a chance to take a look yet?

**Comment by ibrierley** [comment]
Hiya and sorry, been a bit distracted, I did just take a quick look. I think the overall concept looks good, and the first version worked well. Second/main version, the "Big" line doesn't seem to work right, it seems to sometimes draw in one go, sometimes not (as though it keeps getting reset early?). Small one works well.

The only other thing I had been pondering (and got stuck), was if there was a way to do it without needing to replicate some of the existing painter code and reuse that (so if that updates, yours does too). But I couldn't quickly see an elegant way, and often get a bit stuck with that with flutter_map, and have had to do similar in the past!

I was also wondering, but I think more complex (possibly), is if actually the raw supplied points could be interpolated (and inbetween points calculated), then the existing poly widgets can be used), but I think thats maybe getting too much coding for the project.

**Comment by 1075dn** [comment]
@ibrierley Thanks for taking a peek.
In the second version, what's happening is that the map doesn't repaint the polylines with the updated lengths unless you move the map, besides for a little bit right after you start the animation. The small one only _seems_ to work since its animation duration is short enough that the map repaints most of it right after you start the animation (if you look closely you'll see it doesn't paint the entire line until you move the map).

So really the problem is that the map isn't repainting the lines when the `portionToShow` is changed by the animation.

As for your suggestion of creating a new polyline from an existing one and a `portionToShow`, I started working on it (it's not so complicated if you use the `maps_toolkit` package). While doing that, I realized that it wouldn't be accurate, since I calculated the distance in meters, not in projected pixels, and therefore it would have looked "off" for very long lines. If we want to calculate it according to projected pixels in order for it to be accurate, we would need to project it to pixels, calculate the distance in pixels, and then unproject the resulting polyline portion back to `List<LatLng>`. Then the `PolylinePainter` would project it back to pixels in order to render it on the screen. Do you think it's worth it to do all that?

(Sorry for the long, possibly-hard-to-understand explanation 😉🤷‍♂️)

**Comment by ibrierley** [comment]
Not quite sure I follow the accuracy thing. Meters would be pretty accurate ?

I suppose I was thinking something like add up all the distances from latlongs, figure the timing segments to switch between each point to interpolate against. So point 1->2 interpolate for 0->0.2 seconds, point 2-3 interpolate for 0.2->0.9secs etc depending on the distance between them. Not quite sure if I'm misunderstanding, or maybe we're talking at cross purposes.I was wondering whether something like https://github.com/flutter-studio/animated_interpolation would do it

**Comment by 1075dn** [comment]
Hey @ibrierley,

I hope this explains it better (please excuse my awful video editing skills!):

Say you have a long polyline, stretching from east of Russia to Australia.
![map_line](https://user-images.githubusercontent.com/66363175/118376049-4c202e80-b5ce-11eb-9e0e-a7ca79b88fa9.png)

Now, that polyline across the earth really looks like this (side view to make the curve visible):
![world_line](https://user-images.githubusercontent.com/66363175/118376078-77a31900-b5ce-11eb-8a9a-5dcecc8d2ce8.png)

So if you'd animate the polyline based on its length in meters, it would look like this:
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/66363175/118376116-b933c400-b5ce-11eb-8bd6-1b56e188e3a9.gif)

But depending on your map projection, in your flattened map it would end up with an easing effect like this:
![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/66363175/118376371-44fa2000-b5d0-11eb-838f-d1fed1172ef5.gif)

However, if you'd calculate the animation in projected pixels (the flattened map rendered on screen, not in percentage of distance in meters which would animate it across the earth's sphere), you'd get an even, linear animation:
![ezgif com-gif-maker](https://user-images.githubusercontent.com/66363175/118376237-60186000-b5cf-11eb-887d-e192879db125.gif)

Please let me know if you understand, or if I'm mistaken (as I very well may be!).

Thanks!

**Comment by 1075dn** [comment]
> As for your suggestion of creating a new polyline from an existing one and a `portionToShow`, I started working on it (it's not so complicated if you use the `maps_toolkit` package).

Here's the **(untested)** code I was talking about:
```dart
import 'package:latlong/latlong.dart';
import 'package:maps_toolkit/maps_toolkit.dart' as toolkit;

toolkit.LatLng _convert(LatLng point) {
  return toolkit.LatLng(point.latitude, point.longitude);
}

LatLng _convertBack(toolkit.LatLng point) {
  return LatLng(point.latitude, point.longitude);
}

void calculatePortionOfPolyline(List<LatLng> polyline, double portion) {
  var ourList = polyline.map((e) => _convert(e)).toList();
  var totalDistance = toolkit.SphericalUtil.computeLength(ourList);
  var newDistance = totalDistance * portion;
  List<LatLng> newList = [polyline[0]];
  var currentDistance = 0.0;
  var lastIndex;
  for (lastIndex = 1; lastIndex < ourList.length; lastIndex++) {
    var distanceFromPreviousToCurrent =
        toolkit.SphericalUtil.computeDistanceBetween(
            ourList[lastIndex - 1], ourList[lastIndex]);
    if (currentDistance + distanceFromPreviousToCurrent > newDistance) break;
    currentDistance += distanceFromPreviousToCurrent;
    newList.add(polyline[lastIndex]);
  }
  var distanceLeft = totalDistance - currentDistance;

  if (distanceLeft > 0) {
    var lastPoint = ourList[lastIndex - 1];
    var thisPoint = ourList[lastIndex];
    var heading = toolkit.SphericalUtil.computeHeading(lastPoint, thisPoint);
    var newPoint =
        toolkit.SphericalUtil.computeOffset(lastPoint, distanceLeft, heading);
    newList.add(_convertBack(newPoint));
  }
}
```

@ibrierley Does it make sense to you?

**Comment by ibrierley** [comment]
Yes, I think so, I hadn't realised you were looking at such large overall distances, but the problem now makes sense.

**Comment by 1075dn** [comment]
> The only other thing I had been pondering (and got stuck), was if there was a way to do it without needing to replicate some of the existing painter code and reuse that (so if that updates, yours does too).

#### So we have two options:
- We can either copy the existing painter and add the 'polyline portion' calculation code in there (what my current code does), **_or_**
- We can project the polyline to pixels, calculate the distance in pixels, unproject the resulting polyline portion back to List<LatLng>, and supply that to the existing painter (which would project it back to pixels in order to render it on the screen).

Each has pros and cons:
| | Calculation in painter | Calculation outside painter |
| --- | --- | --- |
| **Updating** | Doesn't update when animation interpolation value changes | Updates when animation interpolation value changes |
| **Performance** | Calculates new path from polyline and portion every time animation interpolation value changes [which is not trivial](https://api.flutter.dev/flutter/dart-ui/Path/computeMetrics.html) | Projects to pixels, unprojects back to `LatLng`, and (painter) projects again back to pixels every time animation interpolation value changes |
| **Polyline length** | Does not allow us to access polyline lengths outside painter | Might (this is useful for setting animation duration in proportion to longest polyline) |

What's your opinion?

**Comment by ibrierley** [comment]
Tbh as its mainly for your orienteering app, I'd go with whichever is the easier to go with now. You pretty much have option 1 ready I think, and fits fine with your use case ? I'm always a bit of way of folk getting distracted on the main thing they want to achieve (i.e orienteering app vs line animating :D).

**Comment by 1075dn** [comment]
So if we're staying with what I have now, do you have any idea as to how to get it to update?

**Comment by ibrierley** [comment]
Just had a play more what I was originally thinking, which doesn't need a plugin or rewrite of any code, will fit straight in and should work as is. It would just need shoving in a class and widget and tidying etc. So more like method 2) I've used a haversine calculation for distance, but you could add any func instead at myDistanceFunc to project and back or whatever. 

Change interpolateBetweenPoints to true/false depending if you just want your discrete points or not.

https://gist.github.com/ibrierley/29878f6513c107410f1899bbd4edfaca

**Comment by 1075dn** [comment]
Thanks so much! I was wondering whether there was a way to avoid looping through the points until the new line had the correct distance, every time the animation value changed - and you nailed it! I'll try to work on integrating that...

On a different note, if I end up projecting the points to get accurate distances, do you think it's a good idea to make the `PolylineLayer` accept the polyline point lists as projected points (as opposed to `LatLng`s), to take out the unnecessary unprojection (to supply to the existing `PolylineLayer`), and subsequent projection again (by the painter for rendering)?

**Comment by ibrierley** [comment]
No problem! I didn't get time to go through it thoroughly, so you may want to just keep an eye the first and last points aren't missed, things like that. I couldn't quite figure out if it was necessary to store the percentages or not, but somehow it felt more natural to do it like that.

Hmm the projection thing. Something feels wrong with needing to change PolylineLayer at all ?

Thinking a bit abstract, should it even need to know if it's a polyline, a polygon, a pixel, or a piece of string ?  That way if someone wants to implement a FlashingFluffledLinePointFilledCurverThing that takes a set of points, it would still work ?

**Comment by 1075dn** [comment]
I pushed new code (based on yours). It can definitely be cleaned up, but it works.
Also it would be nice to have an easier way to animate the polylines, as well as allow for animating multiple polylines.
Any ideas?

**Comment by ibrierley** [comment]
Btw I just added a repo at https://github.com/ibrierley/line_animator/ which is just the interpolator and an included widget padded out a bit if anyone wants. It has your getData file as an example, let me know if you don't want it in there and I'll dig out another set of points.

I won't have chance to update the readme properly until after the holidays here, and I haven't had chance to test it all. Will try and get a video/gif up. Atm it draws a polyline line with an aeroplane following the line changing heading It may be a bit different to your thingy, so I've kept it separate for the moment, and not as a plugin (just needs LatLngs). 

The widget I don't think will handle multiple polylines atm ( can look at that next week maybe), but could call multiple PointInterpolator() with a mapcontroller etc. 

As mentioned, I'll try and get the readme done properly when I get time.
